src/data/fin/tuple.lean

```diff
@@ -154,17 +156,116 @@ begin
       rw [cons_succ, hi] } }
 end
 
-/-- `fin.append ho u v` appends two vectors of lengths `m` and `n` to produce
-one of length `o = m + n`.  `ho` provides control of definitional equality
-for the vector length. -/
-def append {α : Type*} {o : ℕ} (ho : o = m + n) (u : fin m → α) (v : fin n → α) : fin o → α :=
-λ i, if h : (i : ℕ) < m
-  then u ⟨i, h⟩
-  else v ⟨(i : ℕ) - m, (tsub_lt_iff_left (le_of_not_lt h)).2 (ho ▸ i.property)⟩
-
-@[simp] lemma fin_append_apply_zero {α : Type*} {o : ℕ} (ho : (o + 1) = (m + 1) + n)
-  (u : fin (m + 1) → α) (v : fin n → α) :
-  fin.append ho u v 0 = u 0 := rfl
+section append
+
+/-- Append a tuple of length `m` to a tuple of length `n` to get a tuple of length `m + n`.
+This is a non-dependent version of `fin.add_cases`. -/
+def append {α : Type*} (a : fin m → α) (b : fin n → α) : fin (m + n) → α :=
+@fin.add_cases _ _ (λ _, α) a b
+
+@[simp] lemma append_left {α : Type*} (u : fin m → α) (v : fin n → α) (i : fin m) :
+  append u v (fin.cast_add n i) = u i :=
+add_cases_left _ _ _
+
+@[simp] lemma append_right {α : Type*} (u : fin m → α) (v : fin n → α) (i : fin n) :
+  append u v (nat_add m i) = v i :=
+add_cases_right _ _ _
+
+lemma append_right_nil {α : Type*} (u : fin m → α) (v : fin n → α) (hv : n = 0) :
+  append u v = u ∘ fin.cast (by rw [hv, add_zero]) :=
+begin
+  refine funext (fin.add_cases (λ l, _) (λ r, _)),
+  { rw [append_left, function.comp_apply],
+    refine congr_arg u (fin.ext _),
+    simp },
+  { have := fin.cast hv r,
+    exact this.elim0 }
+end
+
+@[simp] lemma append_elim0 {α : Type*} (u : fin m → α) :
+  append u fin.elim0 = u ∘ fin.cast (add_zero _) :=
+append_right_nil _ _ rfl
+
+lemma append_left_nil {α : Type*} (u : fin m → α) (v : fin n → α) (hu : m = 0) :
+  append u v = v ∘ fin.cast (by rw [hu, zero_add]) :=
+begin
+  refine funext (fin.add_cases (λ l, _) (λ r, _)),
+  { have := fin.cast hu l,
+    exact this.elim0 },
+  { rw [append_right, function.comp_apply],
+    refine congr_arg v (fin.ext _),
+    simp [hu] },
+end
+
+@[simp] lemma elim0_append {α : Type*} (v : fin n → α) :
+  append (fin.elim0 : _ → α) v = v ∘ fin.cast (zero_add _) :=
+append_left_nil _ _ rfl
+
+lemma append_assoc {p : ℕ} {α : Type*} (a : fin m → α) (b : fin n → α) (c : fin p → α) :
+  append (append a b) c = append a (append b c) ∘ fin.cast (add_assoc _ _ _) :=
+begin
+  ext i,
+  rw function.comp_apply,
+  apply fin.add_cases (λ l, _) (λ r, _) i,
+  { rw append_left,
+    apply fin.add_cases (λ ll, _) (λ lr, _) l,
+    { rw append_left,
+      simp [cast_add_cast_add] },
+    { rw append_right,
+      simp [cast_add_nat_add], }, },
+  { rw [append_right],
+    simp [←nat_add_nat_add] },
+end
```

Comment:
For consistency:
```suggestion
begin
  ext i,
  rw function.comp_apply,
  apply fin.add_cases (λ l, _) (λ r, _) i,
  { rw append_left,
    apply fin.add_cases (λ ll, _) (λ lr, _) l,
    { rw append_left,
      simp [cast_add_cast_add] },
    { rw append_right,
      simp [cast_add_nat_add] , },
  { rw append_right,
    simp [←nat_add_nat_add] },
end
```

I assume there's a reason that the `rw append_left` is not incorporated into `simp`?

Reply:
Did you mean to remove a `}` there? That no longer works.

Reply:
Removed the `[]`s as requested.